### PR TITLE
Wcs join metadata

### PR DESCRIPTION
**Pull request by didntpay**

### CODE REVIEW

**Comment by didntpay** [comment]
Note
The sky boundary query still has the issues of if the center point exists outside of the search box but it overlaps with some of it, that wcs's metadata won't be returned

File: trail/query/tests.py [hunk 1]
```diff
@@ -73,3 +73,10 @@ def testBasicWcsQuery(self):
         response = self.metadataDao.queryBySpecifiedSky(queryParam)
 
         self.assertEqual(len(response), 1)
+
+    def testBasicJoinQuery(self):
+        """Tests getMeatadataInSpecifiedSky in metadataDao is functional"""
+        paramDict = {"raLow": 0, "raHigh": 1000, "decLow": 0, "decHigh": 1000, "processor_name__icontains": "fits"}
+        response = self.metadataDao.queryJoinWcsAndMetadataParam(paramDict)
+
+        self.assertEqual(len(response), 1)
```
Comment (mrawls) [inline trail/query/tests.py]: I think this is outdated and you can keep the version from the `main` branch

File: trail/query/views.py [hunk 2]
```diff
@@ -78,6 +78,50 @@ class MetadataDAO(APIView):
...
+        Parameters
+        ----------
+        raLow: float
+            the lower bound for ra
+        raHigh: float
+            the upper bound for ra
+        decLow: float
+            the lower bound for dec
+        decHigh: float
+            the upper bound for dec
```
Comment (mrawls) [inline trail/query/views.py]: These are not parameters, the only parameter that goes into this method is `paramDict`. Documenting these is a great idea though! They should go in a class docstring up around line 73 instead.

File: trail/query/views.py [hunk 3]
```diff
@@ -78,6 +78,50 @@ class MetadataDAO(APIView):
...
+        metadataParams: key-value pairs
```
Comment (mrawls) [inline trail/query/views.py]: Great description! The argument name is `paramDict` however, not `metadataParams`, please update here

File: trail/query/views.py [hunk 4]
```diff
@@ -142,6 +186,33 @@ def queryBySpecifiedSky(self, paramDict):
...
+    # Helpers -----------------------------------------
+    def makeJoinQueryForReverseLookup(self, skyBoundary, paramDict):
+        """
+        Helper function for formatting the queryDict so that it could lookups that span relationship.
+
+        Parameters
+        ----------
+        skyBoundary: The dict that contains all wcs query info
+        paramDict: the unprocessed paramDict which contains queryInfo from join query
+        """
```
Comment (mrawls) [inline trail/query/views.py]: Please add "Returns" section in docstring, describing what kind of thing `result` is and what is in it

File: trail/query/views.py [hunk 5]
```diff
@@ -142,6 +186,33 @@ def queryBySpecifiedSky(self, paramDict):
...
+            if not self.isWcsParam(key):
+                result["metadata__" + key] = val
+
+        result.update(skyBoundary)
+
+        return result
+
+    def isWcsParam(self, key):
+        """
+        Helper function for verifying if the given key is a wcs query
```
Comment (mrawls) [inline trail/query/views.py]: Please add `key` to Parameters section in docstring, and describe that a boolean (True if it's an RA or Dec upper or lower limit, False otherwise) is returned in Returns section

File: trail/query/views.py [hunk 6]
```diff
@@ -272,3 +344,56 @@ def get(self, request):
...
+            the lower bound for ra
+        raHigh: float
+            the upper bound for ra
+        decLow: float
+            the lower bound for dec
+        decHigh: float
+            the upper bound for dec
+        metadataParams: key-value pairs
+            the key-value pairs in the paramDict that contains the criterias for metadata object
+                Ex.(id = 3, processName=fits)
```
Comment (mrawls) [inline trail/query/views.py]: As before, the only parameter documented here should be `request`, but it's a great idea to describe what is in `request`

File: trail/query/views.py [hunk 7]
```diff
@@ -272,3 +344,56 @@ def get(self, request):
...
+
+        Returns
+        -------
+        results: list of metadata objects
+            a list of metadata objects that matches the param specified by users and is in the boundary box.
+
+        Notes
+        -----
+        If users do not specify a dict, the method will return all metadatas.
+        If users do not specify a params, it will return an empty list.
```
Comment (mrawls) [inline trail/query/views.py]: I'm confused by these notes, I don't immediately see how users specify either thing directly to this method. I guess whatever `request` is should contain a dict as well as whatever params is? This seems like a good test coverage opportunity.

**Review by mrawls — COMMENTED** [review]
Just docstring improvement requests. I defer to @DinoBektesevic for final merge approval.

File: trail/query/tests.py (hunk 1 above)
Comment (DinoBektesevic) [reply trail/query/tests.py:82]: Not sure what you mean by the "version from the `main` branch"? I think this tests the newly added functionality where one can query Metadata table on WCS conditionals.

**Review by DinoBektesevic — CHANGES_REQUESTED** [review]
I cloned the PR to test it locally and then noticed some comments that were marked as resolved from the previous PR were left unchanged as well. There was either a misunderstanding and you merged without implementing what comments requested you to do or there was an accident with git rebase/merge/branch/pull somewhere. For example [like this](https://github.com/dirac-institute/trailblazer/pull/46#discussion_r876444210]). 

I think it would be fastest we meet up and go through the code to fix the current and previous PR together.

File: trail/query/coord_conversion.py [hunk 8]
```diff
@@ -9,6 +10,7 @@ def getXYZFromWcs(self, ra, dec):
 
     return {"x": x, "y": y, "z": z}
 
+
 def isWcsQueryParamMissing(self, queryParams):
```
Comment (DinoBektesevic) [inline trail/query/coord_conversion.py]: This doesn't belong here. Probably an leftover artifact from a copy paste.

File: trail/query/coord_conversion.py [hunk 9]
```diff
@@ -1,6 +1,7 @@
-import numpy
+import numpy as np
 
-def getXYZFromWcs(self, ra, dec):
+
+def getXYZFromWcs(ra, dec):
```
Comment (DinoBektesevic) [inline trail/query/coord_conversion.py]: Could live with a docstring, and maybe a better name like `eq2unitsphere` or `equatorial_to_unitsphere` or.... 
Should describe it requires radians.

File: trail/query/rest_views.py [hunk 10]
```diff
@@ -1,24 +1,45 @@
...
 
-from drf_yasg.utils import swagger_auto_schema
-
 
 class GenericTableView(mixins.ListModelMixin, APIView):
+    """
+     GenericTableView that serves as an interface to provide one unify rest api query
+     to different tables
+    """
+
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: Should probably describe it requires specifying `queryset` and `serializer_class` class attributes for the table view to work.

File: trail/query/rest_views.py [hunk 11]
```diff
@@ -1,24 +1,45 @@
...
     queryset = None
     serializer_class = None
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: Should be documented like:
```
queryset = None
"""Initial queryset on which the table queries will subsequently be executed"""
serializer_class = None
"""Appropriate serializer to serialize the returned data."""

File: trail/query/rest_views.py [hunk 12]
```diff
@@ -1,24 +1,45 @@
...
 
-    def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
+    def get_shared_non_query_keys(self, queryDict):
+        """
+        Extracts the keys and values not directly related to the query from the given query dict,
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: So if the first line is not longer than 120 chars required to trigger flake8 (https://github.com/dirac-institute/trailblazer/blob/main/setup.cfg#L5) why are the rest kept so short? Can you get them to look more like 120char wide block?

File: trail/query/rest_views.py [hunk 13]
```diff
@@ -28,35 +49,118 @@ def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
 
         return queryDict, serializerParams, querysetParams
 
-    def get_non_query_keys(self, queryDict, *args, **kwargs):
+    def get_non_query_keys(self, queryDict):
+        """
+        Remove the keys and values not directly related to the queryDict.
+        and copies them into a queryset related parameter
+        and serializer related parameter dictionaries.
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:60]: Should probably specify that it's a no-op by default and that it allows for individual views to override this method to implement specialized behaviour based on keys found in the request.

File: trail/query/rest_views.py [hunk 14]
```diff
@@ -82,9 +186,17 @@ class MetadataView(GenericTableView):
     queryset = Metadata.objects.all()
     serializer_class = MetadataSerializer
 
-    def get_non_query_keys(self, queryDict, *args, **kwargs):
+    def get_non_query_keys(self, queryDict):
+        """
+            Follows the same logic as GenericTableView.get_non_query_keys
+
+        """
+
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:198]: if you want the docs to be inherited you can't define them again here. If you leave this blank they should just be inherited.

File: trail/query/rest_views.py [hunk 15]
```diff
@@ -96,25 +208,95 @@ def get_non_query_keys(self, queryDict, *args, **kwargs):
...
+        """
         if getWcs:
             return self.queryset.prefetch_related("wcs")
         return self.queryset
 
     def is_sky_region_query(self, qparams):
+        """
+            Helper function for describing if the query is a query base on
+            sky area.
+        """
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:242]: You can shorten this to `Returns True if any of the sky region parameters are found in qparams`, Otherwise you have to define all of the sections.

File: trail/query/rest_views.py [hunk 16]
```diff
@@ -96,25 +208,95 @@ def get_non_query_keys(self, queryDict, *args, **kwargs):
...
     def query(self, qparams, querysetParams):
+        """
+        Query the metadata table with the given query params
+
+        Parameters
+        ----------
+        qparams: Dictionary
+            The cleaned user given query dictionary
+        querysetParams: Dictionary
+            The extracted pre-processing indicator fields dictionary
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]:
```
qparams : `dict`
    Dictionary of query parameters.
qurysetParams : `dict`
    Dictionary of parameters passed to initial queryset (see `get_queryset`)

Returns
-----------
queryset : `django.db.models.query.QuerySet`
    Query results

Notes
-------
If nothing is specified in the qparam, the method return all metadata by default.

The currently accepted list of all query related parameters consist of 
columns of `Metadata`, such as instrument and filter, and additional 
parameters that can specify a sky region `raLow`, `raHigh`, `decLow`
and `decHigh`. See `upload.models.Metadata`
```

File: trail/query/rest_views.py [hunk 17]
```diff
@@ -96,25 +208,95 @@ def get_non_query_keys(self, queryDict, *args, **kwargs):
...
+
+
+class WcsView(GenericTableView):
+    queryset = Wcs.objects.all()
+    serializer_class = WcsSerializer
+
+    def get_queryset(self, *args, **kwargs):
+        """
+            Follows the same logic as MetadataView.get_queryset
+        """
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:287]: Remove if you want to inherit otherwise it's reimplemented.

File: trail/query/rest_views.py [hunk 18]
```diff
@@ -96,25 +208,95 @@ def get_non_query_keys(self, queryDict, *args, **kwargs):
...
+        return self.queryset
+
+    def get_non_query_keys(self, queryDict, *args, **kwargs):
+        """
+            No op
+        """
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:293]: Remove if you want to inherit otherwise it's reimplemented

File: trail/query/rest_views.py [hunk 19]
```diff
@@ -96,25 +208,95 @@ def get_non_query_keys(self, queryDict, *args, **kwargs):
...
+        return queryDict, {}, {}
+
+    def query(self,  qparams, querysetParams):
+        """
+            Follows the same logic as MetadataView.query
+        """
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:299]: Remove if you want to inherit otherwise it's reimplemented

File: trail/query/rest_views.py [hunk 20]
```diff
@@ -1,24 +1,45 @@
...
+        such as fields or casesensitive
+
+        Parameters
+        ----------
+        queryDict: django.QueryDict
+            The queryDict converted from user inputs
+
+        Returns
+        -------
+        serializerParams : Dictionary
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:40]: dictionary --> `dict`

File: trail/query/rest_views.py [hunk 21]
```diff
@@ -1,24 +1,45 @@
...
+            The dictionary containing all serializer related param specified by each table.
+        querysetParams : Dictionary
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:42]: dictionary --> `dict`

In general, you put the same names here as you would if you were in a python repl. So if you would declare a new dictionary with `dict()` you use dict, if you use a list then it's `list` if you use some class from some other module then it's `module.submodule.class`

File: trail/query/rest_views.py [hunk 22]
```diff
@@ -28,35 +49,118 @@ def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
...
         return queryDict, serializerParams, querysetParams
 
     def bad_request(self, message):
+        """
+        Helper for constructing a django.http.HttpResponseBadRequest
+        with given message
+
+        Parameters
+        ----------
+        message : String
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:129]: `str`

File: trail/query/rest_views.py [hunk 23]
```diff
@@ -28,35 +49,118 @@ def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
...
 
-    def get(self, request, *args, **kwargs):
+    def get(self, request):
+        """
+        Generic get method for REST api, given request.params, it queries
+        on the given querySet and serialize result back to the user.
+
+        Parameters
+        ----------
+        request: Http.request
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:155]: I would guess it's `django.http.request`or something?

File: trail/query/rest_views.py [hunk 24]
```diff
@@ -28,35 +49,118 @@ def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
...
+            The http reuqest sent.
+
+        Returns
+        -------
+        response: Http.Response
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:160]: `django.http.HttpResponse` probably.

File: trail/query/rest_views.py [hunk 25]
```diff
@@ -96,25 +208,95 @@ def get_non_query_keys(self, queryDict, *args, **kwargs):
...
 
         return queryDict, serializerParams, querysetParams
 
     def get_queryset(self, getWcs=False, *args, **kwargs):
+        """
+        Return the queryset for metatable table
+
+        Parameters
+        ----------
+        getWcs: boolean
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:225]: `bool`

File: trail/query/rest_views.py [hunk 26]
```diff
@@ -96,25 +208,95 @@ def get_non_query_keys(self, queryDict, *args, **kwargs):
...
         return any(["raLow" in qparams, "raHigh" in qparams,
                     "decLow" in qparams, "decHigh" in qparams])
 
     def query(self, qparams, querysetParams):
+        """
+        Query the metadata table with the given query params
+
+        Parameters
+        ----------
+        qparams: Dictionary
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: `dict`

File: trail/query/rest_views.py [hunk 27]
```diff
@@ -96,25 +208,95 @@ def get_non_query_keys(self, queryDict, *args, **kwargs):
...
+            The cleaned user given query dictionary
+        querysetParams: Dictionary
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: `dict`

File: trail/query/rest_views.py [hunk 28]
```diff
@@ -96,25 +208,95 @@ def get_non_query_keys(self, queryDict, *args, **kwargs):
...
+            The extracted pre-processing indicator fields dictionary
+
+        Returns
+        -------
+        queryset: Django QuerySet
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:259]: `django.db.models.query.QuerySet`

File: trail/query/rest_views.py [hunk 29]
```diff
@@ -28,35 +49,118 @@ def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
...
+        a dict for serializers relate keys, and a dict for preprocess related keys
+
+        Parameters
+        ----------
+        queryDict: django.QueryDict
+            The queryDict converted from user inputs
+
+        Returns
+        -------
+        queryDict : Dictionary
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: `dict`

File: trail/query/rest_views.py [hunk 30]
```diff
@@ -28,35 +49,118 @@ def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
...
+            The cleaned dictionary that only has fields related to the query for
+            the table.
+        serializerParams : Dictionary
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: `dict`

File: trail/query/rest_views.py [hunk 31]
```diff
@@ -28,35 +49,118 @@ def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
...
+            The dictionary containing all serializerParams that is not related to the query,
+            but meant for the serializers.
+        querysetParams : Dictionary
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: `dict`

File: trail/query/rest_views.py [hunk 32]
```diff
@@ -28,35 +49,118 @@ def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
...
         return queryDict, {}, {}
 
     def extract_params(self, queryDict):
+        """
+        Remove and split fields in queryDict into a dict for only query related keys,
+        a dict for serializers relate keys, and a dict for preprocess related keys
+
+        Parameters
+        ----------
+        queryDict: django.QueryDict
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py:93]: `django.http.QueryDict`

File: trail/query/rest_views.py [hunk 33]
```diff
@@ -28,35 +49,118 @@ def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
...
+        and serializer related parameter dictionaries.
+
+        Parameters
+        ----------
+        queryDict: django.QueryDict
+            The queryDict converted from user inputs
+
+        Returns
+        -------
+        queryDict : Dictionary
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: `dict`

File: trail/query/rest_views.py [hunk 34]
```diff
@@ -28,35 +49,118 @@ def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
...
+            The cleaned dictionary that only has fields related to the query for
+            the table.
+        serializerParams : Dictionary
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: dict

File: trail/query/rest_views.py [hunk 35]
```diff
@@ -28,35 +49,118 @@ def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
...
+            The dictionary containing all serializerParams that is not related to the query,
+            but meant for the serializers.
+        querysetParams : Dictionary
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: dict

File: trail/query/rest_views.py [hunk 36]
```diff
@@ -28,35 +49,118 @@ def get_shared_non_query_keys(self, queryDict, *args, **kwargs):
...
-    def get_non_query_keys(self, queryDict, *args, **kwargs):
+    def get_non_query_keys(self, queryDict):
+        """
+        Remove the keys and values not directly related to the queryDict.
+        and copies them into a queryset related parameter
+        and serializer related parameter dictionaries.
+
+        Parameters
+        ----------
+        queryDict: django.QueryDict
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: `django.http.QueryDict`

File: trail/query/rest_views.py [hunk 37]
```diff
@@ -1,24 +1,45 @@
...
+        """
+        Extracts the keys and values not directly related to the query from the given query dict,
+        and copies them into a queryset related parameter
+        and serializer related parameter dictionaries.
+        Extracts only those keys and values that are shared among all table views,
+        such as fields or casesensitive
+
+        Parameters
+        ----------
+        queryDict: django.QueryDict
```
Comment (DinoBektesevic) [inline trail/query/rest_views.py]: `django.http.QueryDict`

**Review by DinoBektesevic — APPROVED** [review]
Minor fixes mostly on types in the docstring. The rest looks good.
